File: examples/hello-secure-agent-ts/agent.ts
/**
 * hello-secure-agent-ts
 *
 * Demonstrates Clawdstrike policy enforcement with the OpenAI Agents SDK
 * tool shape. Run with --dry-run to see allow/deny decisions without an
 * OpenAI API key.
 *
 * Usage:
 *   npx tsx agent.ts --dry-run
 */

import { Clawdstrike, type Decision } from "@clawdstrike/sdk";
import * as fs from "node:fs";
import * as path from "node:path";

// ---------------------------------------------------------------------------
// Parse flags
// ---------------------------------------------------------------------------

const dryRun = process.argv.includes("--dry-run");

// ---------------------------------------------------------------------------
// Load policy and create session
// ---------------------------------------------------------------------------

const policyPath = path.resolve(import.meta.dirname ?? ".", "policy.yaml");
const cs = await Clawdstrike.fromPolicy(policyPath);
const session = cs.session({ agentId: "hello-secure-agent" });

// ---------------------------------------------------------------------------
// Define tools with inline security checks
// (Same shape as OpenAI Agents SDK function tools)
// ---------------------------------------------------------------------------

const tools = {
  read_file: {
    execute: async (input: { path: string }): Promise<string> => {
      const decision = await session.checkFile(input.path, "read");
      if (decision.status === "deny") {
        return `BLOCKED by ${decision.guard}: ${decision.message}`;
      }
      return fs.readFileSync(input.path, "utf-8");
    },
  },

  write_file: {
    execute: async (input: { path: string; content: string }): Promise<string> => {
      const decision = await session.checkFile(input.path, "write");
      if (decision.status === "deny") {
        return `BLOCKED by ${decision.guard}: ${decision.message}`;
      }
      fs.writeFileSync(input.path, input.content, "utf-8");
      return `Wrote ${input.content.length} bytes to ${input.path}`;
    },
  },

  fetch_url: {
    execute: async (input: { url: string }): Promise<string> => {
      const decision = await session.checkNetwork(input.url);
      if (decision.status === "deny") {
        return `BLOCKED by ${decision.guard}: ${decision.message}`;
      }
      const resp = await fetch(input.url);
      return (await resp.text()).slice(0, 500);
    },
  },
};

// ---------------------------------------------------------------------------
// Dry-run scenarios
// ---------------------------------------------------------------------------

const scenarios = [
  { label: "Read allowed file", fn: () => tools.read_file.execute({ path: "/tmp/workspace/notes.txt" }) },
  { label: "Read /etc/shadow", fn: () => tools.read_file.execute({ path: "/etc/shadow" }) },
  { label: "Write to /tmp/workspace/out.txt", fn: () => tools.write_file.execute({ path: "/tmp/workspace/out.txt", content: "agent output" }) },
  { label: "Write to ~/.ssh/evil_key", fn: () => tools.write_file.execute({ path: `${process.env.HOME}/.ssh/evil_key`, content: "hacked" }) },
  { label: "Fetch api.openai.com", fn: () => tools.fetch_url.execute({ url: "https://api.openai.com" }) },
  { label: "Fetch evil.com", fn: () => tools.fetch_url.execute({ url: "https://evil.com" }) },
];

async function runDryRun(): Promise<void> {
  console.log("=== Clawdstrike Security Demo (dry-run) ===\n");

  // Ensure the sandbox directory and test file exist
  fs.mkdirSync("/tmp/workspace", { recursive: true });
  fs.writeFileSync("/tmp/workspace/notes.txt", "Hello from the secure agent!\n");

  for (const { label, fn } of scenarios) {
    console.log(`Scenario: ${label}`);
    try {
      const result = await fn();
      const firstLine = result.split("\n")[0]?.slice(0, 80) ?? "";
      console.log(`  Result: ${firstLine}\n`);
    } catch (err) {
      console.log(`  Error: ${(err as Error).message}\n`);
    }
  }

  // Print session summary
  const summary = session.getSummary();
  console.log("=== Session Summary ===");
  console.log(`  Session ID:      ${summary.sessionId}`);
  console.log(`  Total checks:    ${summary.checkCount}`);
  console.log(`  Allowed:         ${summary.allowCount}`);
  console.log(`  Warnings:        ${summary.warnCount}`);
  console.log(`  Denied:          ${summary.denyCount}`);
  if (summary.blockedActions.length > 0) {
    console.log(`  Blocked actions: ${summary.blockedActions.join(", ")}`);
  }
}

async function runAgent(): Promise<void> {
  console.log("Agent mode requires an OPENAI_API_KEY. Use --dry-run to demo policy checks.");
  console.log("In a real integration you would pass these tools to the OpenAI Agents SDK.\n");
  await runDryRun();
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

if (dryRun) {
  await runDryRun();
} else {
  await runAgent();
}

File: examples/hello-secure-agent-py/agent.py
"""Secure AI agent demo using Clawdstrike + OpenAI Agents SDK.

Run in dry-run mode (no API key required):
    python agent.py --dry-run

Run with a live OpenAI agent (requires OPENAI_API_KEY):
    python agent.py
"""

from __future__ import annotations

import os
import subprocess
import sys
import urllib.parse

from clawdstrike import Clawdstrike


def main() -> None:
    dry_run = "--dry-run" in sys.argv

    # Load policy from the YAML file next to this script.
    script_dir = os.path.dirname(os.path.abspath(__file__))
    policy_path = os.path.join(script_dir, "policy.yaml")

    cs = Clawdstrike.from_policy(policy_path)
    session = cs.session(agent_id="hello-secure-agent")

    # -- Tool definitions with inline security checks -------------------------

    def read_file(path: str) -> str:
        """Read a file from disk after checking the security policy."""
        decision = session.check_file(path)
        if decision.denied:
            return f"BLOCKED by {decision.guard}: {decision.message}"
        try:
            with open(path) as f:
                return f.read()
        except OSError as exc:
            return f"OS error: {exc}"

    def run_command(cmd: str) -> str:
        """Run a shell command after checking the security policy."""
        decision = session.check_command(cmd)
        if decision.denied:
            return f"BLOCKED by {decision.guard}: {decision.message}"
        try:
            return subprocess.check_output(cmd, shell=True, text=True, timeout=10)
        except subprocess.CalledProcessError as exc:
            return f"Command failed (exit {exc.returncode}): {exc.output}"
        except subprocess.TimeoutExpired:
            return "Command timed out"

    def fetch_url(url: str) -> str:
        """Fetch a URL after checking the security policy."""
        host = urllib.parse.urlparse(url).hostname or ""
        decision = session.check_network(host)
        if decision.denied:
            return f"BLOCKED by {decision.guard}: {decision.message}"
        import requests  # noqa: E402  (lazy import -- only needed when allowed)

        try:
            resp = requests.get(url, timeout=10)
            return resp.text[:500]
        except requests.RequestException as exc:
            return f"Request error: {exc}"

    # -- Run ------------------------------------------------------------------

    if dry_run:
        _run_dry(read_file, run_command, fetch_url)
    else:
        _run_agent(read_file, run_command, fetch_url)

    # Print session summary regardless of mode.
    summary = session.get_summary()
    print("\n=== Session Summary ===")
    print(f"Total checks: {summary.check_count}")
    print(f"Allowed:      {summary.allow_count}")
    print(f"Denied:       {summary.deny_count}")
    if summary.blocked_actions:
        print(f"Blocked:      {summary.blocked_actions}")


def _run_dry(
    read_file,
    run_command,
    fetch_url,
) -> None:
    """Run six demo scenarios without an API key."""
    print("=== Clawdstrike Security Demo (dry-run) ===\n")

    # Set up a workspace file for the allowed-read scenario.
    os.makedirs("/tmp/workspace", exist_ok=True)
    with open("/tmp/workspace/notes.txt", "w") as f:
        f.write("Hello from the secure agent!")

    scenarios: list[tuple[str, object]] = [
        ("Read allowed file (/tmp/workspace/notes.txt)", lambda: read_file("/tmp/workspace/notes.txt")),
        ("Read blocked file (/etc/shadow)",              lambda: read_file("/etc/shadow")),
        ("Run allowed command (ls -la /tmp)",             lambda: run_command("ls -la /tmp")),
        ("Run blocked command (rm -rf /)",               lambda: run_command("rm -rf /")),
        ("Fetch allowed host (api.openai.com)",          lambda: fetch_url("https://api.openai.com")),
        ("Fetch blocked host (evil.com)",                lambda: fetch_url("https://evil.com")),
    ]

    for name, fn in scenarios:
        print(f"Scenario: {name}")
        result = fn()
        first_line = result.splitlines()[0] if result else "(empty)"
        print(f"  Result: {first_line}\n")


def _run_agent(
    read_file,
    run_command,
    fetch_url,
) -> None:
    """Run a live agent with the OpenAI Agents SDK."""
    try:
        from agents import Agent, Runner, function_tool  # type: ignore[import-untyped]
    except ImportError:
        print(
            "ERROR: openai-agents is not installed.\n"
            "Install it with:  pip install openai-agents\n"
            "Or run in dry-run mode:  python agent.py --dry-run",
            file=sys.stderr,
        )
        sys.exit(1)

    read_file_tool = function_tool(read_file)
    run_command_tool = function_tool(run_command)
    fetch_url_tool = function_tool(fetch_url)

    agent = Agent(
        name="secure-assistant",
        instructions=(
            "You are a helpful assistant with secure tool access. "
            "Try to help the user while respecting security policies."
        ),
        tools=[read_file_tool, run_command_tool, fetch_url_tool],
    )

    result = Runner.run_sync(agent, "List files in /tmp and read /tmp/workspace/notes.txt")
    print(result.final_output)


if __name__ == "__main__":
    main()
